### ISSUE TITLE
Found an issue about time zone

### ISSUE THREAD

**Issue by sqc1999** [rxtur/BlogEngine.NET#42]
When trying to create a post, the `Post.DateCreate` property is set to the local time, but when saving it to the database, `BlogSettings.ToUtc` uses the server side time zone. There is also some similar problems in other parts of the project.
I think some time conversion should be added between server side code and client side code. Maybe adding them to the api controllers is a good idea.
Sorry for my bad English.

**Comment by rxtur** [comment]
It should save to UTC on server and use offset to display on the client. 
Some details here: http://www.rtur.net/blog/post/2015/09/04/crazy-fun-times

Do you also have issue with published posts not showing for anonymous users for some time?

**Comment by sqc1999** [comment]
I don't think it is working correctly.
if I visit `editpost.cshtml`, `$scope.post.dateCreated`is set to the client side time using `moment.format()` in `posteditorController`, and then if I click the save button, it directly calls `PostsController.Post()` with the client side time sent, and then in `BlogProvier.InsertPost()` it is converted using the server side time zone which is set in blog settings.
I could not find anything about client side time zone.

**Comment by sqc1999** [comment]
And `BlogSettings.ToUtc` is trying to convert client side time to utc using `DateTime.ToUniversalTime()` which uses server side time zone.

**Comment by sqc1999** [comment]
Err...so what's the meaning of `Server time offset` in the blog settings? Is it the difference between (server time and utc) or (client time and server time).
Most code int the project reflects the latter, but the translation in labels.zh-cn.resx uses the former(I'm from China).

**Comment by sqc1999** [comment]
And if the setting means "the difference between client time and server time", `BlogSettings.ToUtc` will not working correctly because it is using the server side time zone.

**Comment by alexandrudanpop** [comment]
Hey rxtur, 

I had the problem when I deployed www.control-f5.com (BE 3.1.0.1). My server was hosted in the US, and I'm from Europe. When logged with admin I could see posts, anonymous user could not. I tried accurate time offset in settings but it didn't work for me.

The fix I used:
Step 1. Open Post.cs file from BlogEngine.Core project
Step 2. Find following code
else if (this.IsPublished && this.DateCreated <= DateTime.Now.AddHours(BlogSettings.Instance.Timezone))
Step 3. Replace above code by this
else if (this.IsPublished)
Step 4. Find following code
//return (this.IsPublished && this.DateCreated <= DateTime.Now.AddHours(BlogSettings.Instance.Timezone)) &&
//        this.IsDeleted == false;
Step 5. Replace above code by this
return this.IsPublished && this.IsDeleted == false;

I found this fix on : 
http://www.itorian.com/2015/04/blog-post-not-visible-in-blogengine.html

Hope this helps,
Alex

**Comment by rxtur** [comment]
alexandrudanpop - we know we had timezone issue in 3.1, it was completely rewritten in 3.2.

sqc1999 - we not using server time, we using universal (UTC) time which is the same on server and client. Time offset is calculate client time off the universal time. Problem is with anonymous (not logged in) users that might see "wrong" time on the post, but they still should see published post. Do you have an issue with published post not visible to not logged in user in 3.2?

**Comment by Cambridgeport90** [comment]
The issue is whether that code snippet you gave will fix the issue? I really want to go back to this, but the issue

I have is that I can’t go changing the time zone to a different format on the server, because if I do, then that wrecks Active Directory and the GPOs that are set there.

**Comment by rxtur** [comment]
I think we getting all confused on this one and start digging into code without understanding differences in setup and even app versions.

We not looking into older version, only 3.2.

We need clearly identify problem and provide steps to reproduce.

Like:
1. I'm using 3.2 out of the box (or specify what you changed).
2. I'm in the US central time (-6) and my server in Eastern time (-5).
3. Time zone selected in settings CT (-6).
4. When I create new post and publish, not logged in users can't see it.

Something like this. I need to follow your steps and get same result to fix it.

Currently I'm running 3.2 on five different servers, with 3 hosts (Azure, Arvixe and WinHost) not counting local servers. And none of them have issue with publishing date/time. But there seems to be few people who do have this issue. I need somehow to recreate it and so far I can't.

**Comment by Cambridgeport90** [comment]
My setup is as follows. I've got a web server (it's part of a domain), running BlogEngine 3.2, and the latest version of ASP.net (also runs PHP for other sites on the server). my Framework version is 4.6, and my timezone is Eastern standard time.

**Comment by sqc1999** [comment]
@rxtur 

1. So why does the setting named "server time offset"? I think it should be changed.
2. I don't think it's a good practice to set the time zone in the settings, you can always fetch the client side time zone using javascript code on the client and convert it when it is sent to web api controller. It is not convenient to set the time zone every time when I change the time zone I am in.
**3. `BlogSettings.ToUtc` should be changed because it is using server side time zone to convert time. I think it is exactly what you are looking for.**
4. In `DbBlogProvider.cs` line 1350, `ToUtc` should be changed to `FromUtc`.

**Comment by rxtur** [comment]
Need to understand issue you having and be able to reproduce it. You still didn't explain your problem and I have no idea what you using, what you doing and what is not working.

Again: 
Last thing I want to do is start making code changes hoping it will fix problem I can't see in the first place.

**Comment by Cambridgeport90** [comment]
When I manage to redeploy tomorrow then I’ll let you know as I don’t have it set up at all … chimed in on this because I uninstalled BlogEngine because of this issue. But basically what would happen is that every post after the first one would only show up to logged in users. I’ll test though for the rest of this week and see what happens.

**Comment by rxtur** [comment]
I added this code to troubleshoot the PostView.ascx in the theme.

`    <section class="post-body text">

        <script>
            var d = new Date();
        </script>
        <table>
            <tr>
                <td>Local time</td>
                <td><script>document.write(d.toString())</script></td>
            </tr>
            <tr>
                <td>Server time</td>
                <td><%=DateTime.Now %></td>
            </tr>
            <tr>
                <td>Post created</td>
                <td><%=Post.DateCreated %></td>
            </tr>
            <tr>
                <td>Settings zone</td>
                <td><%=BlogSettings.Instance.TimeZoneId %></td>
            </tr>
            <tr>
                <td>UTC time</td>
                <td><%=DateTime.UtcNow %></td>
            </tr>
            <tr>
                <td>Settings ToUtc</td>
                <td><%=BlogSettings.Instance.ToUtc() %></td>
            </tr>
            <tr>
                <td>Settings FromUtc</td>
                <td><%=BlogSettings.Instance.FromUtc() %></td>
            </tr>
        </table>
        <hr />

        <asp:PlaceHolder ID="BodyContent" runat="server" />
        <%=AdminLinks %>
    </section>`

Then I ran 2 test cases:
1. "test one" with server in UTC +6 (Novosibirsk)
2. "test two" with server in UTC -10 (Hawaii)

Client is -6 (Chicago), time zone also -6 (CST).

This gives results below, with posts published as expected.

![tz](https://cloud.githubusercontent.com/assets/1932785/13802338/bb5f206c-eb08-11e5-9158-3f6dae4fa468.PNG)

So basically this tests publishing to server way ahead of my local time zone and way behind, and both work just fine.

**Comment by Cambridgeport90** [comment]
That should work, but how is that going to affect a web server in a domain who’s group policy settings dictate Eastern Standard time? I see no reason to exclude my web server from the domain and from ease of management due to something as seemingly trivial as this. (unless it’s more than I can understand, being a beginner coder myself?)

**Comment by scrummie02** [comment]
Your server time due to domain membership is irrelevant.  Just use the proper settings for time in the application 





On Wed, Mar 16, 2016 at 6:55 AM -0700, "Katherine Moss" <notifications@github.com> wrote:










That should work, but how is that going to affect a web server in a domain who’s group policy settings dictate Eastern Standard time? I see no reason to exclude my web server from the domain and from ease of management due to something as seemingly trivial as this. (unless it’s more than I can understand, being a beginner coder myself?)

**Comment by rxtur** [comment]
Latest code might fix this, please test if it works for you.

**Comment by Cambridgeport90** [comment]
Definitely will … I actually plan on having my site run from the latest code always … plan to set up a deployment server that poles this site for it and deploys it …
